pkg/sql/schemachanger/sctest/comparator.go

```diff
@@ -0,0 +1,283 @@
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/errors"
+	"github.com/lib/pq"
+	"github.com/stretchr/testify/require"
+)
+
+// This file contains common logic for performing comparator testing
+// between legacy and declarative schema changer.
+
+type StmtLineReader interface {
+	// HasNextLine returns true if there is more lines of SQL statements to process.
+	HasNextLine() bool
+
+	// NextLine retrieve the next line of SQL statements to process.
+	NextLine() string
+}
+
+// CompareLegacyAndDeclarative is the core logic for performing comparator
+// testing between legacy and declarative schema changer.
+// It reads sql statements (mostly DDLs), one by one, from `ss` and execute them
+// in a cluster using legacy schema changer and in another using declarative
+// schema changer. It asserts that, if the statement fails, they must fail with
+// a PG error with the same pg code. If the statement succeeded, all descriptors
+// in the cluster should end up in the same state.
+func CompareLegacyAndDeclarative(t *testing.T, ss StmtLineReader) {
+	ctx := context.Background()
+	var linesExecutedSoFar []string
+	legacyTSI, legacySQLDB, _ := serverutils.StartServer(t, base.TestServerArgs{})
+	declarativeTSI, declarativeSQLDB, _ := serverutils.StartServer(t, base.TestServerArgs{})
+	defer legacyTSI.Stopper().Stop(ctx)
+	defer declarativeTSI.Stopper().Stop(ctx)
+	legacyTDB := sqlutils.MakeSQLRunner(legacySQLDB)
+	declarativeTDB := sqlutils.MakeSQLRunner(declarativeSQLDB)
+	legacyTDB.Exec(t, "SET use_declarative_schema_changer = off;")
+	declarativeTDB.Exec(t, "SET use_declarative_schema_changer = on;")
+
+	// A helper to determine whether table `name`'s current primary key column is
+	// the implicit `rowid` column.
+	isCurrentPrimaryKeyColumnRowID := func(name *tree.UnresolvedObjectName) bool {
+		res := legacyTDB.QueryStr(t, fmt.Sprintf(`SELECT crdb_internal.pb_to_json('cockroach.sql.sqlbase.Descriptor', descriptor) -> 'table' -> 'primaryIndex' ->> 'keyColumnNames' FROM system.descriptor WHERE id = '%v'::REGCLASS;`, name.String()))
+		return res[0][0] == `["rowid"]`
+	}
+
+	// A helper to determine whether the connection is currently in an open
+	// transaction.
+	isInAnOpenTransaction := func() bool {
+		res := legacyTDB.QueryStr(t, `SHOW transaction_status;`)
+		return res[0][0] == `Open`
+	}
+
+	for ss.HasNextLine() {
+		line := ss.NextLine()
+		line = modifyBlacklistedStmt(t, line, isCurrentPrimaryKeyColumnRowID, isInAnOpenTransaction)
+		_, errLegacy := legacySQLDB.Exec(line)
+		if pgcode.MakeCode(string(getPQErrCode(errLegacy))) == pgcode.FeatureNotSupported {
+			continue
+		}
+		_, errDeclarative := declarativeSQLDB.Exec(line)
+		requireNoErrOrSameErrCode(t, line, errLegacy, errDeclarative)
+		linesExecutedSoFar = append(linesExecutedSoFar, line)
+
+		// Perform a descriptor identity check after a DDL.
+		if containsStmtOfType(t, line, tree.TypeDDL) {
```

Comment:
Is there any reason to not perform the metaDataIdentityCheck on every statement?

Reply:
another super nit: `continue`ing if the statement is not a DDL would be much more readable.